// react-controllable/test/components.js
import {createElement, Component, PropTypes} from 'react'


export class Bare extends Component {
  render() {
    return createElement('div')
  }
}


export class ConflictingProperties extends Bare {
  target() {}
}


export class WithFormOptions extends Bare {
  static propTypes = {
    tabIndex: PropTypes.number.isRequired
  }
}


export class WithPropTypes extends Bare {
  static propTypes = {
    onControl: PropTypes.func.isRequired,
    active: PropTypes.bool,
  }
  static defaultProps = {
    className: "defaultClassName",
    active: true,
  }
}


export class WithPrimaryAction extends Component {
  controlPrimaryAction() {}

  render() {
    return (
      createElement('div', this.shell(),
        createElement('button', Object.assign({type: 'button'}, this.focusable()),
          createElement('div', this.target(),
            createElement('div', {}, 'content')
          )
        )
      )
    )
  }
}


export class WithoutPrimaryAction extends Component {
  render() {
    return (
      createElement('div', this.shell(),
        createElement('div', this.target(),
          createElement('input', this.focusable())
        )
      )
    )
  }
}
